### DISCUSSION TITLE
Patterns for E2E testing nested text?

### DISCUSSION THREAD

**Discussion by Poundex** [Instawork/hyperview#986] [Q&A] [answered]
Hi,

I was wondering if anybody had hit up against the same problem as I have and may perhaps have a solution or workaround? I am nesting `<text />` elements to create inline hyperlinks in text. Everything renders OK, and the links are clickable. But when I look at the "native Android DOM" I see that I have just one `<TextView />`:
![Screenshot_20241105_124309](https://github.com/user-attachments/assets/f6fc270d-0a5e-4cc5-93e7-fe6ddd54c214)

I am trying to E2E test this, but have no idea how I can go about clicking the individual links in the text to make sure they behave properly. 

Does anyone have any ideas, please?

Thanks

**Comment by flochtililoch** [comment]
You can set the `id` attribute on your hyperview element (i.e. `text`, `view` etc) - this will [be mapped](https://github.com/Instawork/hyperview/blob/master/src/services/index.ts#L82-L85) to the appropriate prop, depending on the platform, and will help you target the element in your E2E framework.

**Comment by Poundex** [comment]
The outer IDs, and IDs for all other elements are fine and selectable. But when nesting `<text />`, the nested entries don't have a corresponding native element.

For instance the HXML:
```
<text testID="net.poundex.traq:id/timeline_point_description_672a606e99521742eb3d0707" style="itemLabel">
    Testing markers
    <text testID="marker3" href="/log?q=%2Bone" style="marker marker-tag">+one</text> 
    and 
    <text testID="marker4" href="/log?q=%40two" style="marker marker-mention">@two</text> and
    <text testID="marker5" href="/log?q=%3Athree%3Dfour" style="marker marker-property">:three=four</text>
</text>
```

gets rendered as
```
<android.widget.TextView
    resource-id="net.poundex.traq:id/timeline_point_description_672a606e99521742eb3d0707"
    text="Testing markers +one and @two and :three=four" />
```

While the styles and the actions/links of these nested `<text />` entries are displayed properly and function properly, they do not seem to have their own dedicated native element so I was wondering how to select the individual links (the nested texts) inside the outer text?

**Reply by flochtililoch** [reply]
Have you try using the `id` attribute? `testID` is the underlying React Native prop, but it might not be mapped directly in the case of nested texts. I'm not certain this will work too, but worth a shot - let me know if not, it might be a bug.

i..e

```
<text id="net.poundex.traq:id/timeline_point_description_672a606e99521742eb3d0707" style="itemLabel">
    Testing markers
    <text id="marker3" href="/log?q=%2Bone" style="marker marker-tag">+one</text> 
    and 
    <text id="marker4" href="/log?q=%40two" style="marker marker-mention">@two</text> and
    <text id="marker5" href="/log?q=%3Athree%3Dfour" style="marker marker-property">:three=four</text>
</text>

**Comment by Poundex** [comment]
It's exactly the same when just using `id`. Even though everything looks/works right in the app, there's only one native element looking after the entire text, even though some of it is styled and has clickable actions.

**Comment by adamstep** [comment, marked as answer]
I'm not really an expert on Android UI testing, but I found some links indicating this might be a general problem on the Android platform. See this [Stack Overflow question](https://stackoverflow.com/questions/52256409/how-can-i-click-on-text-link-inside-in-textview-in-appium) and the Appium discussion linked from there. It seems like the best (hacky) solution is to match the textview by ID, and then apply coordinate offset to touch the place where your link is. Seems like that would break with dynamic text or different device screen sizes, but I haven't found any better solutions.

**Comment by Poundex** [comment]
I hadn't come across that particular thread. I thought it might be something like that but couldn't find anything confirming it. This is well to the right of Hyperview and RN, then. Appreciate you both having a look, though.
